### PR TITLE
Add file describing how to add or modify specialized families of instructions.

### PR DESCRIPTION
**Pull request by markshannon**
We've had a few contributors wanting to implement specialized instructions.
To help out those contributors, I thought it would be useful add a "how to" describing the process.

### CODE REVIEW

File: Python/adaptive.md [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+This can be tested quickly:
+* `globals->keys->dk_version == expected_version`
+
+and the operation can be performed quickly:
+* `value = globals->keys->entries[index].value`.
+
+Because it is impossible to measure the performance of an instruction without also
+measuring unrelated factors, the assessment of the quality of a specialization will require some judgement.
+
+As a general rule, specialized instructions should be much faster than the base instruction.
```
Comment (Fidget-Spinner) [inline Python/adaptive.md]: Just a question: how `much` faster exactly? Or is this evaluated case-by-case using our own judgement?

**Review by Fidget-Spinner — APPROVED** [review]
LGTM. I noticed some of the lower level implementation details aren't described (like the code and data layout, being positive and negative indexes on a contiguous array). But those will probably change quite a bit over time, so I think it's fine to omit them.

File: Python/adaptive.md (hunk 1 above)
Comment (markshannon) [reply Python/adaptive.md]: It depends, so yes use your judgement.

If a specialization covers a lot of cases and reduces the number of misses by a lot, then a x2 speedup would be fine. For a more narrow specialization, something more like x4 would be appropriate.

I'm talking about the time spent in the instruction only, so for call instructions don't count the time spent in the callee or the overhead of making a C call.

**Review by gvanrossum — COMMENTED** [review]
Can you fold lines longer than 80 characters?

Adding more code examples would make the doc more accessible. At the very least indicate where to find the examples you do mention in the source code.

**Comment by pablogsal** [comment]
I would advice to copy/move this guide to https://github.com/python/devguide as this is going to be much more visible there than a readme in the Python directory.

**Comment by gvanrossum** [comment]
> I would advice to copy/move this guide to https://github.com/python/devguide as this is going to be much more visible there than a readme in the Python directory.

Not clear that we need to give it much visibility though. There are probably only 20-30 opcodes that are candidates for specialization and only very few people (including everyone reading this PR :-) will need it. There are other readmes sprinkled throughout the code for similarly obscure things.

Maybe there should be a mention of this readme in ceval.c or specialize.c though?

**Comment by pablogsal** [comment]
> Maybe there should be a mention of this readme in ceval.c or specialize.c though?

Right, that is probably a much better approach :+1:

File: Python/adaptive.md [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+The core part of PEP 659 (specializing adaptive interpreter) is the families of instructions that perform the adaptive specialization.
+
+A family of instructions has the following fundamental properties:
+
+* It corresponds to a single instruction in the code generated by the bytecode compiler.
+* It has a single adaptive instruction that records an execution count and,
+   at regular intervals, attempts to specialize itself. If not specializing, it executes
+  the non-adaptive instruction.
+* It has at least one specialized form of the instruction that is tailored for a particular value or set of values at runtime.
+* All members of the family have access to same number of cache entries.
```
Comment (iritkatriel) [inline Python/adaptive.md]:
```suggestion
* All members of the family have access to the same number of cache entries.
```

File: Python/adaptive.md [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+Keeping `Nmiss` low means that there should be specializations for almost
+all values seen by the base instruction. Keeping `sum(Ti*Ni)` low means keeping `Ti`
+low which means minimizing branches and dependent memory accesses (pointer chasing).
+These two objectives may be in conflict, requiring judgement and experimentation to
+design the family of instructions.
+
+### Gathering data
+
+Before choosing how to specialize an instruction, it is important to gather some data. What are the pattern of usage of the base instruction?
```
Comment (iritkatriel) [inline Python/adaptive.md]:
```suggestion
Before choosing how to specialize an instruction, it is important to gather some data. What are the patterns of usage of the base instruction?
```

**Review by gvanrossum — APPROVED** [review]
Let's land this!

**Pull request merged**